pkg/materialize/source_table.go

```diff
@@ -0,0 +1,192 @@
+        ON mz_tables.id = mz_mysql_source_tables.id
+    LEFT JOIN mz_internal.mz_postgres_source_tables
+        ON mz_tables.id = mz_postgres_source_tables.id
+    JOIN mz_roles
+        ON mz_tables.owner_id = mz_roles.id
+    LEFT JOIN (
+        SELECT id, comment
+        FROM mz_internal.mz_comments
+        WHERE object_type = 'table'
+        AND object_sub_id IS NULL
+    ) comments
+        ON mz_tables.id = comments.id
+`)
+
+func SourceTableId(conn *sqlx.DB, obj MaterializeObject) (string, error) {
+	p := map[string]string{
+		"mz_tables.name":    obj.Name,
+		"mz_schemas.name":   obj.SchemaName,
+		"mz_databases.name": obj.DatabaseName,
+	}
+	q := sourceTableQuery.QueryPredicate(p)
+
+	var t SourceTableParams
+	if err := conn.Get(&t, q); err != nil {
+		return "", err
+	}
+
+	return t.TableId.String, nil
+}
+
+func ScanSourceTable(conn *sqlx.DB, id string) (SourceTableParams, error) {
+	q := sourceTableQuery.QueryPredicate(map[string]string{"mz_tables.id": id})
+
+	var t SourceTableParams
+	if err := conn.Get(&t, q); err != nil {
+		return t, err
+	}
+
+	return t, nil
+}
+
+type SourceTableBuilder struct {
+	ddl                Builder
+	tableName          string
+	schemaName         string
+	databaseName       string
+	source             IdentifierSchemaStruct
+	upstreamName       string
+	upstreamSchemaName string
+	conn               *sqlx.DB
+}
+
+func NewSourceTableBuilder(conn *sqlx.DB, obj MaterializeObject) *SourceTableBuilder {
+	return &SourceTableBuilder{
+		ddl:          Builder{conn, Table},
+		tableName:    obj.Name,
+		schemaName:   obj.SchemaName,
+		databaseName: obj.DatabaseName,
+		conn:         conn,
+	}
+}
+
+func (b *SourceTableBuilder) QualifiedName() string {
+	return QualifiedName(b.databaseName, b.schemaName, b.tableName)
+}
+
+func (b *SourceTableBuilder) Source(s IdentifierSchemaStruct) *SourceTableBuilder {
+	b.source = s
+	return b
+}
+
+func (b *SourceTableBuilder) UpstreamName(n string) *SourceTableBuilder {
+	b.upstreamName = n
+	return b
+}
+
+func (b *SourceTableBuilder) UpstreamSchemaName(n string) *SourceTableBuilder {
+	b.upstreamSchemaName = n
+	return b
+}
+
+func (b *SourceTableBuilder) Rename(newName string) error {
+	oldName := b.QualifiedName()
+	b.tableName = newName
+	newName = b.QualifiedName()
+	return b.ddl.rename(oldName, newName)
+}
+
+func (b *SourceTableBuilder) Drop() error {
+	qn := b.QualifiedName()
+	return b.ddl.drop(qn)
+}
+
+// BaseCreate provides a template for the Create method
+func (b *SourceTableBuilder) BaseCreate(sourceType string, additionalOptions func() string) error {
+	q := strings.Builder{}
+	q.WriteString(fmt.Sprintf(`CREATE TABLE %s`, b.QualifiedName()))
+	q.WriteString(fmt.Sprintf(` FROM SOURCE %s`, b.source.QualifiedName()))
+	q.WriteString(` (REFERENCE `)
+
+	if b.upstreamSchemaName != "" {
+		q.WriteString(fmt.Sprintf(`%s.`, QuoteIdentifier(b.upstreamSchemaName)))
+	}
+	q.WriteString(QuoteIdentifier(b.upstreamName))
+
+	q.WriteString(")")
```

Comment:
mentioned above that this can be made optional for kafka and single-output load generator sources